### PR TITLE
feat: Locker smart contract

### PR DESCRIPTION
**Pull request by ermvrs**
Includes smart contract that locks LP for specific time passed in constructor.
Also provides two view methods to return how much time left for unlock and how many tokens locked.

Test may be failure because of snforge version. I need help at point please comment out if you understand why declare error caused by. Tested with sn version 0.9.

### CODE REVIEW

File: contracts/src/locker.cairo [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    #[constructor]
+    fn constructor(ref self: ContractState, lock_time: u64) {
+        self.lock_time.write(lock_time);
+    }
+
+    #[external(v0)]
+    impl Locker of super::ILocker<ContractState> {
+        // Locks liquidity in this smart contract.
+        fn lock_liquidity(ref self: ContractState, lp_address: ContractAddress, amount: u256) {
```
Comment (abdelhamidbakhta) [inline contracts/src/locker.cairo:33]: Use checks / effects / interactions pattern.

Example: https://github.com/sablier-labs/v2-core/blob/main/src/SablierV2LockupLinear.sol#L449-L513

File: contracts/src/sample_erc20.cairo [hunk 2]
```diff
@@ -0,0 +1,253 @@
+use starknet::ContractAddress;
+
+#[starknet::interface]
+trait IERC20<TContractState> {
```
Comment (abdelhamidbakhta) [inline contracts/src/sample_erc20.cairo:4]: use Open Zeppelin ERC20 component. You can find an example in their readme here: https://github.com/OpenZeppelin/cairo-contracts?tab=readme-ov-file#using-the-library

File: contracts/tests/test_locker.cairo [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    lp_calldata.append('TEST');
+    lp_calldata.append('TST');
+    lp_calldata.append(18.into());
+    lp_calldata.append(owner_address); // Owner
+    lp_calldata.append(5000.into()); // 5000 tokens
+    lp_calldata.append(0.into());
+
+    let lp_address = lp_contract.deploy(@lp_calldata).unwrap();
+
+    // start_prank(locker_address, owner_address.try_into().unwrap());
```
Comment (abdelhamidbakhta) [inline contracts/tests/test_locker.cairo:36]: remove commented code

File: contracts/tests/test_locker.cairo [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    owner: ContractAddress
+}
+
+fn setup() -> Setup {
+    let owner_address: felt252 =
+        0x123123123123123123123123123;
+    let locker_contract = declare('Locker');
+
+    let mut locker_calldata = ArrayTrait::new();
+    locker_calldata.append(200_u64.into()); // 200ms lock
```
Comment (abdelhamidbakhta) [inline contracts/tests/test_locker.cairo:20]: use mock timestamp to not rely on actual time passed

File: contracts/tests/test_locker.cairo [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    Setup {
+        locker: ILockerDispatcher { contract_address: locker_address },
+        lp_token: IERC20Dispatcher { contract_address: lp_address },
+        owner: owner_address.try_into().unwrap()
+    }
+}
+
+#[test]
+fn test_lock_lp() {
+    let setup = setup();
```
Comment (abdelhamidbakhta) [inline contracts/tests/test_locker.cairo:50]: comment properly all tests and separate into clear steps for each test with comments like:
- Setup
- Mock
- Test
- Assertion

**Review by abdelhamidbakhta — CHANGES_REQUESTED** [review]
left some comments

**Comment by abdelhamidbakhta** [comment]
@enitrat would love if you could review also

File: contracts/src/lib.cairo [hunk 6]
```diff
@@ -1 +1,3 @@
 mod unruggable_memecoin;
+mod locker;
+mod sample_erc20;
```
Comment (enitrat) [inline contracts/src/lib.cairo:3]: missing line before EOF (do fmt checks pass?)

File: contracts/src/locker.cairo [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            assert(amount > 0, 'Zero lock');
```
Comment (enitrat) [inline contracts/src/locker.cairo:34]: amount is of type u256 and can't be negative, prefer an equality check instead

also: can you create an `Errors` module, a `const ZERO_LOCK:felt252 = 'Zero lock`  and use this instead? Same for all the future asserts

```suggestion
            assert(amount != 0, errors::ZERO_LOCK);
```

File: contracts/src/locker.cairo [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            assert(net_receive == amount, 'Transfer fee on LP');
+
+            // Create lock
+            self.locks.write((lpAddress, caller, current_time), amount);
+
+            // Emits event
+        }
+
+        fn unlock_liquidity(ref self: ContractState, lp_address: ContractAddress, lock_time: u64) {
+            assert(lock_time > 0, 'Lock time zero');
```
Comment (enitrat) [inline contracts/src/locker.cairo:79]: equality check + error const (for all asserts)

File: contracts/tests/test_locker.cairo (hunk 4 above)
Comment (ermvrs) [reply contracts/tests/test_locker.cairo:20]: Isnt it mock already?

**Pull request closed without merging**